src/sentry/static/sentry/app/utils/withTeamProjects.tsx

```diff
@@ -0,0 +1,89 @@
+import React from 'react';
+import {isEqual} from 'lodash';
+
+import {Team, Organization, Project} from 'app/types';
+import getDisplayName from 'app/utils/getDisplayName';
+import {Client} from 'app/api';
+
+// We require these props when using this HOC
+type DependentProps = {
+  api: Client;
+  organization: Organization;
+  teams: Team[];
+  loadingTeams: boolean;
+};
+
+type InjectedProjectsProps = {
+  projects: Project[];
+  loadingProjects: boolean;
+};
+
+type State = {
+  projects: Project[];
+  loadingProjects: boolean;
+};
+
+const withTeamProjects = <P extends InjectedProjectsProps>(
+  WrappedComponent: React.ComponentType<P>
+) =>
+  class extends React.Component<
+    Omit<P, keyof InjectedProjectsProps> &
+      Partial<InjectedProjectsProps> &
+      DependentProps,
+    State
+  > {
+    static displayName = `withTeamProjects(${getDisplayName(WrappedComponent)})`;
+
+    state = {
+      projects: [],
+      loadingProjects: true,
+    };
+
+    componentDidMount() {
+      if (!this.props.loadingTeams) {
+        this.fetchProjects();
+      }
+    }
+
+    componentDidUpdate(prevProps) {
+      if (!this.props.loadingTeams && !isEqual(prevProps.teams, this.props.teams)) {
+        this.fetchProjects();
+      }
+    }
+
+    async fetchProjects() {
+      this.setState({
+        loadingProjects: true,
+      });
+      const promises = this.getTeamsProjectPromises();
+      const projects: Project[][] = await Promise.all(promises);
+      this.setState({
+        projects: projects.flat(),
+        loadingProjects: false,
+      });
+    }
+
+    getTeamsProjectPromises() {
+      return this.getTeamsProjectEndpoints().map(endpoint => {
+        return this.props.api.requestPromise(endpoint);
+      });
+    }
+
+    getTeamsProjectEndpoints() {
+      return this.props.teams!.map(team => {
+        return `/teams/${this.props.organization.slug}/${team.slug}/projects/`;
+      });
```

Comment:
Should we consider making an endpoint to fetch all the teams/projects at once? This looks like it could do many requests for accounts with many teams.

Reply:
Yup, in fact the previous endpoint in the withUsersTeams HOC actually has the project information so this withTeamProjects HOC is actually unnecessary. Will update this PR after I merge in the individual HOC https://github.com/getsentry/sentry/pull/14859